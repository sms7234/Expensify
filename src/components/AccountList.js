import React from  'react';
import {connect} from 'react-redux';
import AccountListItem from './AccountListItem';
import selectAccounts from '../selectors/accounts';


export const AccountList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Accounts</div>
      <div className="show-for-desktop">Accounts</div>
      <div className="show-for-desktop">Description</div>
    </div>
    <div className="list-body">
    {
      props.accounts.length === 0 ? (
        <div className="list-item--message">
          <span>No accounts to show</span>
        </div>
      ): (
        props.accounts.map((item) => {return <AccountListItem key={item.id}{...item}/>;})
      )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    accounts: selectAccounts(state.accounts, state.filters)
  };
};

export default connect(mapStateToProps)(AccountList);
