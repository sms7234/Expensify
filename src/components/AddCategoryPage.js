import React from 'react';
import CategoryForm from './CategoryForm';
import {connect} from 'react-redux';
import {startAddCategory} from '../actions/categories';

export class AddCategoryPage extends React.Component{
  onSubmit=(category) => {
    this.props.startAddCategory(category);
    this.props.history.push('/categories');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Category </h1>
          </div>
        </div>
        <div className="content-container">
          <CategoryForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddCategory:(category) => dispatch(startAddCategory(category))
});

export default connect(undefined, mapDispatchToProps)(AddCategoryPage);
