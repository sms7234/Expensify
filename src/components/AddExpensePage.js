import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect} from 'react-redux';
import startAddExpense from '../actions/expenses';

export class AddExpensePage extends React.Component{
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/expenses');
  };
  render(){
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit = {this.onSubmit}
            categoryList = {this.props.categoryList}
          />
        </div>
      </div>
    );
  }
};

const mapStateToProps=(state) => {
  return {
    categoryList: state.categories.map((item)=>{return({label: item.category, value: item.category})})
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);
