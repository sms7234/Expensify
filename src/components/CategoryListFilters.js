import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter} from '../actions/filters';

export class CategoryListFilters extends React.Component {
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  render () {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="search categories"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CategoryListFilters);
