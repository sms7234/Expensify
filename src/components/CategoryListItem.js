import React from 'react';
import {Link} from 'react-router-dom';

const CategoryListItem=({dispatch, id, category, description}) =>(
  <Link className="list-item" to={`/add`}>
    <div>
      <h3 className="list-item__title">{category}</h3>
    </div>
    <h3 className="list-item__data">{description}</h3>
  </Link>
)

export default CategoryListItem;
