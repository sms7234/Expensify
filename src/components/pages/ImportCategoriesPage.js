import React from 'react';
import {connect} from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import {startAddCategory} from '../../actions/categories';
import ImportInstructions from '../independents/ImportInstructions';
import ImportUploadCategories from '../independents/ImportUploadCategories';
import ImportCategoryListItem from '../items/ImportCategoryListItem';
import ImportSummary from '../independents/ImportSummary';


export class ImportCategoriesPage extends React.Component {
  state= {
    data: [],
    validation: [],
    buttonSave: true,
    buttonCheck: true,
  };
  onDataUpload = (data, validation) => {
    this.setState(()=>({data, validation}));
    this.setState(()=>({
      buttonUpload: true,
      buttonCheck: false
    }));
  };
  onDataCheck = () => {
    const val= this.state.validation;
    this.state.data.forEach((item,index) => {
      if(item.Category===""){
        val[index]=false;
      } else {
        val[index]=true;
      }
    });
    this.setState(()=>({validation: val}))
    if (val.includes(false) || val.includes(null)) {
      this.setState(()=>({
        buttonSave:true,
      }))
    } else {
      this.setState(()=>({buttonSave:false}))
    }
  };
  onDataAdd = () =>{
    const newData = {
      Category: '',
      Description: ''
    };
    const data = this.state.data;
    const validation = this.state.validation;
    data.push(newData);
    validation.push(null);
    this.setState(()=>({data, validation, buttonCheck: false}))
  };
  onItemRemove = (index) => {
    const data = this.state.data;
    const validation = this.state.validation;
    data.splice(index,1);
    validation.splice(index,1);
    this.setState(()=>({data, validation}));
  };
  onItemSave=(update, index) => {
    const data = this.state.data;
    data[index] = update;
    this.setState(()=>({data}));
  };
  onSubmit=() =>{
    const convertedData = [];
    this.state.data.forEach((item)=>{
      convertedData.push({
        category: item.Category,
        description: item.Description
      })
    });
    convertedData.forEach((item) => {
      this.props.startAddCategory(item)
    })
    this.props.history.push('/categories');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title"> Import Categories </h1>
            <h2 className="page-header__subtitle">Manually input or import csv data to add multiple categories simultaneously</h2>
            <ImportSummary className="page-header__subtitle" qty={this.state.data} wordSingular={'category'} wordPlural={'categories'}/>
          </div>
        </div>
        <ImportInstructions
          dataFormat={['Category', 'Description']}
        />
        <ImportUploadCategories
          onDataUpload={this.onDataUpload}
          data={this.state.data}
          validation={this.state.validation}
        />
        <div className="content-container">
          <div className="list-header">
            <h4 className="show-for-mobile">Category List </h4>
            <h4 className="show-for-desktop">Category</h4>
            <h4 className="show-for-desktop">Description</h4>
          </div>
          <div className="list-body">
          {
            !this.state.data ? (
              <div className="list-item--message">
                <span>No Data Available</span>
              </div>
            ): (
              this.state.data.map((item,index) => {
                return <ImportCategoryListItem
                  key = {uuidv4()}
                  {...item}
                  id={index}
                  validation={this.state.validation[index]}
                  onRemove={this.onItemRemove}
                  onSave={this.onItemSave}
                />;
              })
            )
          }
          </div>
          <div className="content-container--buttons">
            <button className="button" onClick={this.onDataAdd}>Add Data</button>
            <button className="button--check" disabled={this.state.buttonCheck} onClick={this.onDataCheck}>Check Data</button>
            <button className="button" disabled={this.state.buttonSave} onClick={this.onSubmit}>Submit Data</button>
          </div>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category))
});

export default connect(undefined, mapDispatchToProps)(ImportCategoriesPage);
