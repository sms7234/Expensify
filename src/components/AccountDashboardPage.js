import React from 'react';
import AccountList from './AccountList';
import AccountSummary from './AccountSummary';
import AccountListFilters from './AccountListFilters';

const AccountDashboardPage = () => (
  <div>
    <AccountSummary />
    <AccountListFilters />
    <AccountList />
  </div>
)

export default AccountDashboardPage;
