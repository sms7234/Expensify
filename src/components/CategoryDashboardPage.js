import React from 'react';
import CategoryList from './CategoryList';
import CategoryListFilters from './CategoryListFilters';
import CategorySummary from './CategorySummary';

const CategoryDashboardPage = () => (
  <div>
    <CategorySummary />
    <CategoryListFilters />
    <CategoryList />
  </div>
)

export default CategoryDashboardPage;
