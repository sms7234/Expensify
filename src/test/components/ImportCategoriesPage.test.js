import React from 'react';
import {shallow} from 'enzyme';
import {ImportCategoriesPage} from '../../components/ImportCategoriesPage';

const fakeData = [
  {
    Category: 'Groceries',
    Description: ''
  },
  {
    Category: 'Healthcare',
    Description: 'blah'
  }
];

const altData1=[
  {
    Category: '',
    Description: ''
  }
];


test('render page with no data',() => {
  const wrapper = shallow(<ImportCategoriesPage />);
  expect(wrapper).toMatchSnapshot();
});

test('test Add data button', () => {
  const wrapper = shallow(<ImportCategoriesPage />);
  wrapper.find('button').at(0).simulate('click');
  expect(wrapper.state('data')).toEqual([{
    Category: '',
    Description: ''
  }]);
});

test('test check button with correct data', () => {
  const wrapper = shallow(<ImportCategoriesPage />);
  wrapper.setState({data: fakeData});
  wrapper.setState({validation: [null, null]})
  wrapper.find('button').at(1).simulate('click');
  expect(wrapper.state('buttonSave')).toBe(false);
  const validation = wrapper.state('validation');
  expect(validation.includes(false)).toBe(false);
});

test('test check button with blank imported fields', () => {
  const wrapper = shallow(<ImportCategoriesPage />);
  wrapper.setState({data: altData1});
  wrapper.setState({validation: [null]})
  wrapper.find('button').at(1).simulate('click');
  expect(wrapper.state('buttonSave')).toBe(true);
  const validation = wrapper.state('validation');
  expect(validation.includes(false)).toBe(true);
})

test('check that submit function runs', () => {
  const onSubmit = jest.fn()
  const history = {push: jest.fn()};
  const wrapper = shallow(<ImportCategoriesPage startAddCategory={onSubmit} history={history}/>);
  wrapper.setState({data: fakeData});
  wrapper.find('button').at(2).simulate('click');
  expect(onSubmit).toHaveBeenCalled();
  expect(history.push).toHaveBeenLastCalledWith('/categories');
})
