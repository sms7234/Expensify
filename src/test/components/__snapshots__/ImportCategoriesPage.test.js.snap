// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render page with no data 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
         Import Categories 
      </h1>
      <h2
        className="page-header__subtitle"
      >
        Manually input or import csv data to add multiple categories simultaneously
      </h2>
      <ImportSummary
        className="page-header__subtitle"
        qty={Array []}
        wordPlural="categories"
        wordSingular="category"
      />
    </div>
  </div>
  <_default
    dataFormat={
      Array [
        "Category",
        "Description",
      ]
    }
  />
  <ImportUploadCategories
    data={Array []}
    onDataUpload={[Function]}
    validation={Array []}
  />
  <div
    className="content-container"
  >
    <div
      className="list-header"
    >
      <h4
        className="show-for-mobile"
      >
        Category List 
      </h4>
      <h4
        className="show-for-desktop"
      >
        Category
      </h4>
      <h4
        className="show-for-desktop"
      >
        Description
      </h4>
    </div>
    <div
      className="list-body"
    />
    <div
      className="content-container--buttons"
    >
      <button
        className="button"
        onClick={[Function]}
      >
        Add Data
      </button>
      <button
        className="button--check"
        disabled={true}
        onClick={[Function]}
      >
        Check Data
      </button>
      <button
        className="button"
        disabled={true}
        onClick={[Function]}
      >
        Submit Data
      </button>
    </div>
  </div>
</div>
`;
